packages/tools/scripts/rewrite-internal-redirects/index.ts

```diff
@@ -0,0 +1,307 @@
+    return visit(tree, (node: Node & { url?: string }) => {
+      if (!is(node, 'link') && !is(node, 'definition')) return
+
+      // Only check internal links
+      if (node.url && !node.url.startsWith('#') && isInternalUrl(node.url)) {
+        const urlToRedirect = node.url.startsWith('/')
+          ? node.url
+          : new URL(node.url).pathname
+
+        let redirectUrl
+
+        if (isInternalUrl(node.url, product)) {
+          redirectUrl = checkAndApplyRedirect(urlToRedirect, redirects)
+        } else {
+          const [, hash] = node.url.split('#')
+          redirectUrl = fetch(node.url, { method: 'HEAD' }).then(
+            (res) =>
+              `${res.url}${hash && !res.url.includes('#') ? `#${hash}` : ''}`
+          )
+        }
+
+        if (redirectUrl) {
+          const data = file.data
+          if (!data.internalRedirects) data.internalRedirects = []
+
+          data.internalRedirects.push({
+            source: node.url,
+            destination: redirectUrl,
+          })
+
+          if (!(redirectUrl instanceof Promise)) {
+            node.url = redirectUrl
+          }
+        }
+      }
+    })
+  }
+}
+
+// from: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Regular_Expressions#escaping
+function escapeRegExp(string) {
+  return string.replace(/[.*+?^${}()|[\]\\]/g, '\\$&') // $& means the whole matched string
+}
+
+/**
+ * In order to preserve the structure of the content
+ */
+function applyRedirectToContent(content, { source, destination }) {
+  // matches definitions OR links
+  const pattern = new RegExp(
+    `(^\\[.*\\]: )${escapeRegExp(source)}((?:#.*)?$)|(\\]\\()${escapeRegExp(
+      source
+    )}((?:#.*)?\\))`,
+    'gim'
+  )
+
+  return content.replace(
+    pattern,
+    (_match, definitionPrefix, definitionSuffix, linkPrefix, linkSuffix) => {
+      if (definitionPrefix && typeof definitionSuffix !== 'undefined') {
+        return `${definitionPrefix}${destination}${definitionSuffix}`
+      }
+
+      if (linkPrefix && linkSuffix) {
+        return `${linkPrefix}${destination}${linkSuffix}`
+      }
+    }
+  )
+}
+
+interface Data {
+  internalRedirects: { source: string; destination: string }[]
+}
+type ParsedFile = Partial<VFile> & { data?: Data }
+
+export default async function main(product: string) {
+  const contentFiles: ParsedFile[] = []
+
+  let dirPath = path.join(cwd, 'content')
+  if (product === 'tutorials') {
+    dirPath = path.join(cwd, 'content', 'tutorials')
+  }
+
+  for await (const entry of readdirp(dirPath, {
+    directoryFilter: ['!partials'],
+    fileFilter: ['*.md', '*.mdx'],
+  })) {
+    const fileContents = await fs.promises.readFile(entry.fullPath, 'utf-8')
+
+    contentFiles.push({
+      raw: String(fileContents),
+      contents: String(fileContents),
+      path: entry.fullPath,
+    })
+  }
+
+  let redirects = []
+  if (product !== 'tutorials') {
+    redirects = loadRedirects()
+  }
```

Comment:
@thiskevinwang this is where we might also handle `terraform`. We will not attempt to load the redirects file.